java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth/ServerAuthMiddleware.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.arrow.flight.CallInfo;
+import org.apache.arrow.flight.CallStatus;
+import org.apache.arrow.flight.FlightServerMiddleware;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Middleware that's used to validate credentials during the handshake and verify
+ * the bearer token in subsequent requests.
+ */
+public class ServerAuthMiddleware implements FlightServerMiddleware {
+  private static final Logger logger = LoggerFactory.getLogger(ServerAuthMiddleware.class);
+
+  /**
+   * Factory for accessing ServerAuthMiddleware.
+   */
+  public static class Factory implements FlightServerMiddleware.Factory<ServerAuthMiddleware> {
+    private final ServerAuthHandler authHandler;
+    private final GeneratedBearerTokenAuthHandler bearerTokenAuthHandler;
+
+    /**
+     * Construct a factory with the given auth handler.
+     * @param authHandler The auth handler what will be used for authenticating requests.
+     */
+    public Factory(ServerAuthHandler authHandler) {
+      this.authHandler = authHandler;
+      bearerTokenAuthHandler = authHandler.enableCachedCredentials() ?
+          new GeneratedBearerTokenAuthHandler() : null;
+    }
+
+    @Override
+    public ServerAuthMiddleware onCallStarted(CallInfo callInfo, CallHeaders incomingHeaders, CallContext context) {
+      logger.debug("Call name: {}", callInfo.method().name());
+      // Check if bearer token auth is being used, and if we've enabled use of server-generated
+      // bearer tokens.
+      if (authHandler.enableCachedCredentials()) {
+        final String bearerTokenFromHeaders =
+            AuthUtilities.getValueFromAuthHeader(incomingHeaders, AuthConstants.BEARER_PREFIX);
+        if (bearerTokenFromHeaders != null) {
+          final ServerAuthHandler.AuthResult result = bearerTokenAuthHandler.authenticate(incomingHeaders);
+          context.put(AuthConstants.PEER_IDENTITY_KEY, result.getPeerIdentity());
+          return new ServerAuthMiddleware(result.getPeerIdentity(), result.getBearerToken().get());
+        }
+      }
+
+      // Delegate to server auth handler to do the validation.
+      final ServerAuthHandler.AuthResult result = authHandler.authenticate(incomingHeaders);
+      final String bearerToken;
+      if (authHandler.enableCachedCredentials()) {
+        bearerToken = bearerTokenAuthHandler.registerBearer(result);
+      } else {
+        bearerToken = result.getBearerToken().get();
+      }
+      context.put(AuthConstants.PEER_IDENTITY_KEY, result.getPeerIdentity());
+      return new ServerAuthMiddleware(result.getPeerIdentity(), bearerToken);
+    }
+  }
+
+  private final String bearerToken;
+  private final String peerIdentity;
+
+  public ServerAuthMiddleware(String peerIdentity, String bearerToken) {
+    this.peerIdentity = peerIdentity;
+    this.bearerToken = bearerToken;
+  }
+
+  @Override
+  public void onBeforeSendingHeaders(CallHeaders outgoingHeaders) {
+    if (bearerToken != null &&
```

Comment:
The same goes for outbound headers - an auth handler should be able to send whatever outbound headers it might need, instead of hardcoding a particular header here. SPNEGO is an example of an auth scheme that uses WWW-Negotiate instead of Authorization in responses, for example.

Reply:
Done. The logic in what to write, if anything, for outbound headers is now part of the AuthResult interface.